scripts/metadata_comparison/metadata_comparison/extractor.py

```diff
@@ -0,0 +1,219 @@
+    else:
+        msg = f'Invalid Cromwell URL {value}. Expected {url_regex.pattern}'
+        raise argparse.ArgumentTypeError(msg)
+
+
+def gcs_path_regex_validator(value):
+    """
+    Validates then extracts the bucket and object-path from a GS string. Returned as a pair.
+    eg:
+        'gs://bucket/path/to/directory/' -> ('bucket', 'path/to/directory')
+    """
+    gcs_regex = re.compile('^gs://([a-zA-Z0-9-]+)/(([a-zA-Z0-9-]+/)*[a-zA-Z0-9-]+)/?$')
+    m = gcs_regex.match(value)
+    if m:
+        return m.group(1), m.group(2)
+    else:
+        msg = f'Invalid GCS path {value}. Expected {gcs_regex.pattern}'
+        raise argparse.ArgumentTypeError(msg)
+
+
+def get_operation_id_number(value):
+    """
+    Validates then extracts from PAPI operation IDs just the final number.
+    eg:
+        'projects/project_name/operations/01234567891011121314' -> '01234567891011121314'
+    """
+
+    operation_regex = re.compile('^.*/([^/]*)')
+    m = operation_regex.search(value)
+    if m:
+        return m.group(1)
+    else:
+        msg = f'Unexpected operation ID {value}. Expected something like {operation_regex.pattern}'
+        raise Exception(msg)
+
+
+def uploadLocalCheckout():
+    #   # Make a snapshot of the local Cromwell git repo
+    #
+    #   hash=$(git rev-parse HEAD)
+    #
+    #   # Make sure all files are checked in.
+    #   if [[ $(git status --short | wc -l) -ne 0 ]]; then echo "stuff needs to be checked in"; exit 1; fi
+    #
+    #   git_snapshots_dir="../git-snapshots"
+    #   mkdir -p "${git_snapshots_dir}"
+    #   git_snapshot="${git_snapshots_dir}/$hash"
+    #   git clone . --local "${git_snapshot}"
+    raise Exception("Not Implemented")
+
+
+def fetch_raw_workflow_metadata(cromwell_url, workflow, verbose):
+    url = f'{cromwell_url}/api/workflows/v1/{workflow}/metadata?expandSubWorkflows=true'
+    if verbose:
+        print('Fetching Cromwell metadata from...', url, end='...')
+    result = requests.get(url)
+    if verbose:
+        print(' fetched!')
+    return result.content, result.json()
+
+
+def find_operation_ids_in_metadata(json_metadata):
+    """Finds all instances of PAPI operations IDs in a workflow, and the API to call to retrieve metadata"""
+    # {
+    #   "calls": {
+    #     "workflow_name.task_name": [
+    #       {
+    #         "backend": "Papi"
+    #         "jobId": "projects/broad-dsde-cromwell-dev/operations/5788303950667477684",
+    papi_operation_to_api_mapping = {}
+
+    calls = json_metadata.get('calls', {})
+    for callname in calls:
+        attempts = calls[callname]
+        for attempt in attempts:
+            jobId = attempt.get('jobId')
+            if jobId:
+                api = 'lifesciences' if 'beta' in attempt.get('backend', '').lower() else 'genomics'
+                papi_operation_to_api_mapping[jobId] = api
+    return papi_operation_to_api_mapping
+
+
+def read_genomics_papiv1_operations_metadata(jobId, api, genomics_client):
```

Comment:
unused

Reply:
My intellij python plugin is clearly not tuned up very high. I wonder if I should be pointing pycharm at this dir while working on python 🤔

Reply:
![Fotolia_31566718_XS](https://user-images.githubusercontent.com/961771/79616884-20352f80-80d4-11ea-9e06-c18008b3db75.jpg)

Reply:
Presumably representing "python charming"...?

Fun fact: reminds me of the time during my first internship when I was still kind of scared and nervous of the whole "workplace environment" thing. I thought I'd be proactive and take a look at this new-to-me language called "python", so I opened up - of course - "`python.com`". Well it turns out that it was (at least at that time, I'll leave it to you to re-verify this now if you want to) an extremely NSFW website. So I quickly closed it down, but spent the whole rest of the day incredibly nervous in case anybody had been watching my screen when it happened.

I guess what I'm saying is... I've personally been burned by the internet before and I'm still even now a little nervous to type "python anything" into a browser. And "python charming" would certainly be quite high up on my do-not-image-search list.

Reply:
https://youtu.be/-o-7MmhqNfA?t=10

(and yes, python charming)